File: chatbot/backend-python/src/rag/embedder.py
"""
Embedding utilities for the RAG pipeline.

Uses sentence-transformers for local embeddings (no API key needed).
Model: all-MiniLM-L6-v2 (384-dim, fast, lightweight ~80MB).

First run will download the model automatically. Subsequent runs use cache.
"""

from typing import Optional

import numpy as np
from sentence_transformers import SentenceTransformer

# Lazy-loaded singleton — model loads on first use, then stays in memory
_model: Optional[SentenceTransformer] = None


def _get_model(model_name: str = "all-MiniLM-L6-v2") -> SentenceTransformer:
    """Get or create the sentence-transformers model singleton."""
    global _model
    if _model is None:
        _model = SentenceTransformer(model_name)
    return _model


def embed_text(text: str, model_name: str = "all-MiniLM-L6-v2") -> list[float]:
    """
    Generate an embedding vector for a single text string.

    Args:
        text: The text to embed.
        model_name: Sentence-transformers model name (default: all-MiniLM-L6-v2).

    Returns:
        A list of floats representing the embedding vector.
    """
    model = _get_model(model_name)
    embedding = model.encode(text, normalize_embeddings=True)
    return embedding.tolist()


def embed_batch(
    texts: list[str],
    model_name: str = "all-MiniLM-L6-v2",
    show_progress: bool = False,
) -> list[list[float]]:
    """
    Generate embeddings for multiple texts in batch.

    More efficient than calling embed_text multiple times.

    Args:
        texts: List of texts to embed.
        model_name: Sentence-transformers model name.
        show_progress: Whether to show a progress bar.

    Returns:
        A list of embedding vectors (each a list of floats).
    """
    model = _get_model(model_name)
    embeddings = model.encode(texts, normalize_embeddings=True, show_progress_bar=show_progress)
    return [e.tolist() for e in embeddings]


def cosine_similarity(vec_a: list[float], vec_b: list[float]) -> float:
    """
    Compute cosine similarity between two vectors.

    Since sentence-transformers normalizes embeddings by default,
    this is simply the dot product of the two vectors.
    """
    if len(vec_a) != len(vec_b):
        raise ValueError(
            f"Vector dimension mismatch: {len(vec_a)} vs {len(vec_b)}"
        )
    return float(np.dot(vec_a, vec_b))


def find_top_k(
    query_vector: list[float],
    chunk_vectors: list[list[float]],
    k: int = 5,
) -> list[dict]:
    """
    Find the top-k most similar chunks to a query vector.

    Args:
        query_vector: The embedding of the query.
        chunk_vectors: List of chunk embedding vectors.
        k: Number of top results to return.

    Returns:
        List of dicts with 'index' and 'score' keys, sorted by score descending.
    """
    scored = [
        {"index": i, "score": cosine_similarity(query_vector, vec)}
        for i, vec in enumerate(chunk_vectors)
    ]
    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:k]

File: chatbot/backend/src/rag/embedder.ts
/**
 * Embedding utilities for the RAG pipeline.
 * Generates embeddings using OpenAI's API and performs cosine similarity search.
 */

import { Env } from '../config'

/**
 * Generate an embedding vector for a single text string.
 * Uses OpenAI's text-embedding-3-small model.
 */
export async function embedText(
  text: string,
  env: Env
): Promise<number[]> {
  const apiKey = env.OPENAI_API_KEY
  if (!apiKey) {
    throw new Error(
      'OPENAI_API_KEY is required for embeddings. Set it via `npx wrangler secret put OPENAI_API_KEY`'
    )
  }

  const model = env.EMBEDDING_MODEL || 'text-embedding-3-small'

  const response = await fetch('https://api.openai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      input: text,
      encoding_format: 'float',
    }),
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`OpenAI Embedding API error (${response.status}): ${error}`)
  }

  const data = (await response.json()) as {
    data: Array<{ embedding: number[] }>
  }
  return data.data[0].embedding
}

/**
 * Generate embeddings for multiple texts in batch.
 * More efficient than calling embedText multiple times.
 */
export async function embedBatch(
  texts: string[],
  env: Env
): Promise<number[][]> {
  const apiKey = env.OPENAI_API_KEY
  if (!apiKey) {
    throw new Error('OPENAI_API_KEY is required for embeddings.')
  }

  const model = env.EMBEDDING_MODEL || 'text-embedding-3-small'

  const response = await fetch('https://api.openai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      input: texts,
      encoding_format: 'float',
    }),
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`OpenAI Embedding API error (${response.status}): ${error}`)
  }

  const data = (await response.json()) as {
    data: Array<{ index: number; embedding: number[] }>
  }
  // Sort by index to maintain original order
  return data.data.sort((a, b) => a.index - b.index).map((d) => d.embedding)
}

/**
 * Compute cosine similarity between two vectors.
 * OpenAI embeddings are normalized to unit length, so this is just the dot product.
 */
export function cosineSimilarity(vecA: number[], vecB: number[]): number {
  if (vecA.length !== vecB.length) {
    throw new Error(
      `Vector dimension mismatch: ${vecA.length} vs ${vecB.length}`
    )
  }

  let dotProduct = 0
  for (let i = 0; i < vecA.length; i++) {
    dotProduct += vecA[i] * vecB[i]
  }
  return dotProduct
}

/**
 * Find the top-k most similar chunks to a query vector.
 * Returns the indices and similarity scores sorted by relevance (highest first).
 */
export function findTopK(
  queryVector: number[],
  chunkVectors: number[][],
  k: number = 5
): Array<{ index: number; score: number }> {
  const scored = chunkVectors.map((vec, index) => ({
    index,
    score: cosineSimilarity(queryVector, vec),
  }))

  // Sort by similarity score descending
  scored.sort((a, b) => b.score - a.score)

  // Return top-k
  return scored.slice(0, k)
}
